ir/reader.go

```diff
@@ -122,26 +132,72 @@ func (self *IrReader) ReadToFunc(r io.Reader, f func(ffi.LogEvent) bool) (ffi.Lo
 	}
 }
 
-// Read the IR stream using the io.Reader until [ffi.LogEvent.Timestamp] >=
-// time. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToEpochTime(r io.Reader, time ffi.EpochTimeMs) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return e.Timestamp >= time })
+// Read the CLP IR stream until a [ffi.LogEventView] is greater than or equal to
+// the given timestamp. Errors are propagated from [ReadToFunc].
+func (self *Reader) ReadToEpochTime(
+	time ffi.EpochTimeMs,
+) (*ffi.LogEventView, error) {
+	return self.ReadToFunc(func(event *ffi.LogEventView) bool { return event.Timestamp >= time })
+}
+
+// Read the CLP IR stream until [strings/Contains] returns true for a
+// [ffi.LogEventView] and the given sub string. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToContains(substr string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.Contains(event.LogMessageView, substr)
+	}
+	return self.ReadToFunc(fn)
+}
+
+// Read the CLP IR stream until [strings/HasPrefix] returns true for a
+// [ffi.LogEventView] and the given prefix. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToPrefix(prefix string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.HasPrefix(event.LogMessageView, prefix)
+	}
+	return self.ReadToFunc(fn)
 }
 
-// Read the IR stream using the io.Reader until [bytes/Contains] returns true
-// for [ffi.LogEvent.Msg] and subslice. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToContains(r io.Reader, subslice []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.Contains(e.Msg, subslice) })
+// Read the CLP IR stream until [strings/HasSuffix] returns true for a
+// [ffi.LogEventView] and the given suffix. Errors are propagated from
+// [ReadToFunc].
+func (self *Reader) ReadToSuffix(suffix string) (*ffi.LogEventView, error) {
+	fn := func(event *ffi.LogEventView) bool {
+		return strings.HasSuffix(event.LogMessageView, suffix)
+	}
+	return self.ReadToFunc(fn)
 }
 
-// Read the IR stream using the io.Reader until [bytes/HasPrefix] returns true
-// for [ffi.LogEvent.Msg] and prefix. Errors are propagated from ReadNextLogEvent.
-func (self *IrReader) ReadToPrefix(r io.Reader, prefix []byte) (ffi.LogEvent, error) {
-	return self.ReadToFunc(r, func(e ffi.LogEvent) bool { return bytes.HasPrefix(e.Msg, prefix) })
+// fillBuf shifts the remaining valid IR in [Reader.buf] to the front and then
+// calls [ioReader.Read] to fill the remainder with more IR. Before reading into
+// the buffer, it is grown by 1.5x if more than 1/4th of it is unconsumed IR.
+// Forwards the return of [ioReader.Read].
```

Comment:
This heuristic seems ok. But in Python libraries (both ffi and loglib), the heuristic we used to grow the buffer is "double the buffer capacity if more than half of the bytes are unconsumed". Any reasons why we choose a different heuristic here? And do we want to be consistent with the Python libraries?